go/consensus/tendermint/apps/beacon/backend_pvss.go

```diff
@@ -0,0 +1,893 @@
+	switch {
+	case height == pvssState.CommitDeadline:
+		ctx.Logger().Debug("OnBeginBlock: height is at commit deadline",
+			"height", height,
+		)
+
+		if pvssState.State != beacon.StateCommit {
+			return fmt.Errorf("beacon: BUG: invalid state: %d (expected %d)", pvssState.State, beacon.StateCommit)
+		}
+
+		// Persist the nodes that failed to commit.
+		var failures []signature.PublicKey
+		for idx, id := range pvssState.Participants {
+			if pvssState.Instance.Commits[idx] == nil {
+				failures = append(failures, id)
+			}
+		}
+		impl.appendFailures(pvssState, failures)
+
+		if ok, totalCommits := pvssState.Instance.MayReveal(); ok {
+			// Update the node status to signify elgibility from the
+			// next epoch.
+			if err := impl.updateNodeStatus(ctx, state, epoch); err != nil {
+				return fmt.Errorf("beacon: failed to update nodes snapshot: %w", err)
+			}
+
+			pvssState.Height = height
+			pvssState.State = beacon.StateReveal
+			if err := state.SetPVSSState(ctx, pvssState); err != nil {
+				return fmt.Errorf("beacon: failed to set updated PVSS state: %w", err)
+			}
+
+			impl.doEmitPVSSEvent(ctx, pvssState)
+		} else {
+			// Round failed: insufficient commits.
+			ctx.Logger().Error("round failed, insufficient commits",
+				"total_commits", totalCommits,
+			)
+
+			return impl.initRound(ctx, state, params, pvssState, pvssState.Epoch)
+		}
+	case height == pvssState.RevealDeadline:
+		ctx.Logger().Debug("OnBeginBlock: height is at reveal deadline",
+			"height", height,
+		)
+
+		if pvssState.State != beacon.StateReveal {
+			return fmt.Errorf("beacon: BUG: invalid state: %d (expected %d)", pvssState.State, beacon.StateReveal)
+		}
+
+		// Persist the nodes that failed to reveal.
+		var failures []signature.PublicKey
+		for idx, id := range pvssState.Participants {
+			if pvssState.Instance.Reveals[idx] == nil {
+				failures = append(failures, id)
+			}
+		}
+		impl.appendFailures(pvssState, failures)
+
+		ok, totalReveals := pvssState.Instance.MayRecover()
+		if ok {
+			// Recover the entropy.
+			var (
+				err      error
+				goodIdxs []int
+			)
+			pvssState.Entropy, goodIdxs, err = pvssState.Instance.Recover()
+			if err != nil {
+				return fmt.Errorf("beacon: failed to recover entropy: %w", err)
+			}
+
+			pvssState.Height = height
+			pvssState.State = beacon.StateComplete
+			if err = state.SetPVSSState(ctx, pvssState); err != nil {
+				return fmt.Errorf("beacon: failed to set updated PVSS state: %w", err)
+			}
+
+			goodParticipants := make(map[int]bool)
+			for _, idx := range goodIdxs {
+				goodParticipants[idx] = true
+			}
+			for idx, id := range pvssState.Participants {
+				switch goodParticipants[idx] {
+				case true:
+					// TODO: Incentivise participation?  Should there be a
+					// reward for participating in the beacon, beyond "time
+					// needs to advance for various rewards to be distributed"?
+				case false:
+					// Slash for failing to participate fully.  A node that
```

Comment:
Does this take into account that the node only gets slashed in case it was able to propose a block during this period?